File: backend/app/routers/workflows.py
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session
from typing import List
from app.database import get_db
from app.models.user import User
from app.models.workflow import Workflow, WorkflowNode, WorkflowConnection
from app.schemas.workflow import (
    WorkflowCreate, WorkflowUpdate, WorkflowResponse, WorkflowWithNodes,
    WorkflowNodeCreate, WorkflowNodeUpdate, WorkflowNodeResponse,
    WorkflowConnectionCreate, WorkflowConnectionUpdate, WorkflowConnectionResponse,
    WorkflowExecutionRequest, WorkflowBulkUpdate, WorkflowPayload, ExecutionResult,
    WorkflowValidationResult
)
from app.auth.dependencies import get_current_active_user
from app.core.tasks import execute_workflow
from app.services.workflow_executor import WorkflowExecutor

router = APIRouter()


@router.get("/workflows", response_model=List[WorkflowResponse])
def get_workflows(
    skip: int = 0,
    limit: int = 100,
    current_user: User = Depends(get_current_active_user),
    db: Session = Depends(get_db)
):
    """Get all workflows for current user"""
    workflows = db.query(Workflow).filter(
        Workflow.owner_id == current_user.id
    ).offset(skip).limit(limit).all()
    return workflows


@router.post("/workflows", response_model=WorkflowResponse)
def create_workflow(
    workflow_data: WorkflowCreate,
    current_user: User = Depends(get_current_active_user),
    db: Session = Depends(get_db)
):
    """Create a new workflow"""
    workflow = Workflow(
        **workflow_data.dict(),
        owner_id=current_user.id
    )
    db.add(workflow)
    db.commit()
    db.refresh(workflow)
    return workflow


@router.get("/workflows/{workflow_id}", response_model=WorkflowWithNodes)
def get_workflow(
    workflow_id: int,
    current_user: User = Depends(get_current_active_user),
    db: Session = Depends(get_db)
):
    """Get a specific workflow with nodes and connections"""
    workflow = db.query(Workflow).filter(
        Workflow.id == workflow_id,
        Workflow.owner_id == current_user.id
    ).first()
    
    if not workflow:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Workflow not found"
        )
    
    return workflow


@router.put("/workflows/{workflow_id}", response_model=WorkflowResponse)
def update_workflow(
    workflow_id: int,
    workflow_data: WorkflowUpdate,
    current_user: User = Depends(get_current_active_user),
    db: Session = Depends(get_db)
):
    """Update a workflow"""
    workflow = db.query(Workflow).filter(
        Workflow.id == workflow_id,
        Workflow.owner_id == current_user.id
    ).first()
    
    if not workflow:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Workflow not found"
        )
    
    for field, value in workflow_data.dict(exclude_unset=True).items():
        setattr(workflow, field, value)
    
    db.commit()
    db.refresh(workflow)
    return workflow


@router.put("/workflows/{workflow_id}/bulk", response_model=WorkflowWithNodes)
def update_workflow_bulk(
    workflow_id: int,
    workflow_data: WorkflowBulkUpdate,
    current_user: User = Depends(get_current_active_user),
    db: Session = Depends(get_db)
):
    """Update a workflow with nodes and connections in a single transaction"""
    workflow = db.query(Workflow).filter(
        Workflow.id == workflow_id,
        Workflow.owner_id == current_user.id
    ).first()
    
    if not workflow:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Workflow not found"
        )
    
    try:
        # Update workflow basic info
        if workflow_data.name is not None:
            workflow.name = workflow_data.name
        if workflow_data.description is not None:
            workflow.description = workflow_data.description
        
        # Handle nodes if provided
        if workflow_data.nodes is not None:
            # Clear existing nodes and connections
            db.query(WorkflowConnection).filter(WorkflowConnection.workflow_id == workflow_id).delete()
            db.query(WorkflowNode).filter(WorkflowNode.workflow_id == workflow_id).delete()
            
            # Create new nodes
            for node_data in workflow_data.nodes:
                node = WorkflowNode(
                    node_id=node_data.id,
                    node_type=node_data.type,
                    name=node_data.data.get('name', 'Unnamed Node'),
                    position_x=node_data.position.get('x', 0),
                    position_y=node_data.position.get('y', 0),
                    config=node_data.data.get('config', {}),
                    workflow_id=workflow_id
                )
                db.add(node)
        
        # Handle edges if provided
        if workflow_data.edges is not None:
            for edge_data in workflow_data.edges:
                connection = WorkflowConnection(
                    connection_id=edge_data.id,
                    source_node_id=edge_data.source,
                    target_node_id=edge_data.target,
                    source_port=edge_data.sourceHandle,
                    target_port=edge_data.targetHandle,
                    condition={"label": edge_data.label} if edge_data.label else None,
                    workflow_id=workflow_id
                )
                db.add(connection)
        
        db.commit()
        db.refresh(workflow)
        return workflow
        
    except Exception as e:
        db.rollback()
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=f"Failed to update workflow: {str(e)}"
        )


@router.post("/workflows/{workflow_id}/validate", response_model=WorkflowValidationResult)
def validate_workflow(
    workflow_id: int,
    workflow_data: WorkflowBulkUpdate,
    current_user: User = Depends(get_current_active_user),
    db: Session = Depends(get_db)
):
    """Validate a workflow structure"""
    workflow = db.query(Workflow).filter(
        Workflow.id == workflow_id,
        Workflow.owner_id == current_user.id
    ).first()
    
    if not workflow:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Workflow not found"
        )
    
    if not workflow_data.nodes or not workflow_data.edges:
        return WorkflowValidationResult(
            valid=False,
            errors=["Workflow must contain nodes and edges"]
        )
    
    # Convert to WorkflowPayload format
    payload = WorkflowPayload(
        id=str(workflow_id),
        name=workflow.name,
        nodes=workflow_data.nodes,
        edges=workflow_data.edges,
        viewport=workflow_data.viewport
    )
    
    # Validate using executor
    executor = WorkflowExecutor()
    validation_result = executor.validate_workflow(payload)
    
    return WorkflowValidationResult(
        valid=validation_result["valid"],
        errors=validation_result["errors"]
    )


@router.delete("/workflows/{workflow_id}")
def delete_workflow(
    workflow_id: int,
    current_user: User = Depends(get_current_active_user),
    db: Session = Depends(get_db)
):
    """Delete a workflow"""
    workflow = db.query(Workflow).filter(
        Workflow.id == workflow_id,
        Workflow.owner_id == current_user.id
    ).first()
    
    if not workflow:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Workflow not found"
        )
    
    db.delete(workflow)
    db.commit()
    return {"message": "Workflow deleted successfully"}


@router.post("/workflows/{workflow_id}/execute")
def execute_workflow_endpoint(
    workflow_id: int,
    execution_request: WorkflowExecutionRequest,
    current_user: User = Depends(get_current_active_user),
    db: Session = Depends(get_db)
):
    """Execute a workflow"""
    workflow = db.query(Workflow).filter(
        Workflow.id == workflow_id,
        Workflow.owner_id == current_user.id
    ).first()
    
    if not workflow:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Workflow not found"
        )
    
    # Start async execution
    task = execute_workflow.delay(
        workflow_id=workflow_id,
        trigger_data=execution_request.trigger_data,
        test_mode=execution_request.test_mode
    )
    
    return {
        "task_id": task.id,
        "status": "executing",
        "message": "Workflow execution started"
    }


@router.post("/workflows/{workflow_id}/test", response_model=ExecutionResult)
def test_workflow_endpoint(
    workflow_id: int,
    execution_request: WorkflowExecutionRequest,
    current_user: User = Depends(get_current_active_user),
    db: Session = Depends(get_db)
):
    """Test execute a workflow synchronously"""
    workflow = db.query(Workflow).filter(
        Workflow.id == workflow_id,
        Workflow.owner_id == current_user.id
    ).first()
    
    if not workflow:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Workflow not found"
        )
    
    # Prepare workflow data for execution
    workflow_data = WorkflowPayload(
        id=str(workflow_id),
        name=workflow.name,
        nodes=[
            {
                "id": node.node_id,
                "type": node.node_type,
                "position": {"x": node.position_x, "y": node.position_y},
                "data": {
                    "name": node.name,
                    "config": node.config or {}
                }
            }
            for node in workflow.nodes
        ],
        edges=[
            {
                "id": conn.connection_id,
                "source": conn.source_node_id,
                "target": conn.target_node_id,
                "sourceHandle": conn.source_port,
                "targetHandle": conn.target_port,
                "label": conn.condition.get("label") if conn.condition else None
            }
            for conn in workflow.connections
        ]
    )
    
    # Execute workflow
    executor = WorkflowExecutor()
    result = executor.execute_workflow(
        workflow_data, 
        trigger_data=execution_request.trigger_data
    )
    
    return result


# Node endpoints
@router.get("/workflows/{workflow_id}/nodes", response_model=List[WorkflowNodeResponse])
def get_workflow_nodes(
    workflow_id: int,
    current_user: User = Depends(get_current_active_user),
    db: Session = Depends(get_db)
):
    """Get all nodes for a workflow"""
    workflow = db.query(Workflow).filter(
        Workflow.id == workflow_id,
        Workflow.owner_id == current_user.id
    ).first()
    
    if not workflow:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Workflow not found"
        )
    
    return workflow.nodes


@router.post("/workflows/{workflow_id}/nodes", response_model=WorkflowNodeResponse)
def create_workflow_node(
    workflow_id: int,
    node_data: WorkflowNodeCreate,
    current_user: User = Depends(get_current_active_user),
    db: Session = Depends(get_db)
):
    """Create a new node in a workflow"""
    workflow = db.query(Workflow).filter(
        Workflow.id == workflow_id,
        Workflow.owner_id == current_user.id
    ).first()
    
    if not workflow:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Workflow not found"
        )
    
    node = WorkflowNode(
        **node_data.dict(),
        workflow_id=workflow_id
    )
    db.add(node)
    db.commit()
    db.refresh(node)
    return node


@router.put("/workflows/{workflow_id}/nodes/{node_id}", response_model=WorkflowNodeResponse)
def update_workflow_node(
    workflow_id: int,
    node_id: int,
    node_data: WorkflowNodeUpdate,
    current_user: User = Depends(get_current_active_user),
    db: Session = Depends(get_db)
):
    """Update a workflow node"""
    node = db.query(WorkflowNode).join(Workflow).filter(
        WorkflowNode.id == node_id,
        Workflow.id == workflow_id,
        Workflow.owner_id == current_user.id
    ).first()
    
    if not node:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Node not found"
        )
    
    for field, value in node_data.dict(exclude_unset=True).items():
        setattr(node, field, value)
    
    db.commit()
    db.refresh(node)
    return node


@router.delete("/workflows/{workflow_id}/nodes/{node_id}")
def delete_workflow_node(
    workflow_id: int,
    node_id: int,
    current_user: User = Depends(get_current_active_user),
    db: Session = Depends(get_db)
):
    """Delete a workflow node"""
    node = db.query(WorkflowNode).join(Workflow).filter(
        WorkflowNode.id == node_id,
        Workflow.id == workflow_id,
        Workflow.owner_id == current_user.id
    ).first()
    
    if not node:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Node not found"
        )
    
    db.delete(node)
    db.commit()
    return {"message": "Node deleted successfully"}


# Connection endpoints
@router.get("/workflows/{workflow_id}/connections", response_model=List[WorkflowConnectionResponse])
def get_workflow_connections(
    workflow_id: int,
    current_user: User = Depends(get_current_active_user),
    db: Session = Depends(get_db)
):
    """Get all connections for a workflow"""
    workflow = db.query(Workflow).filter(
        Workflow.id == workflow_id,
        Workflow.owner_id == current_user.id
    ).first()
    
    if not workflow:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Workflow not found"
        )
    
    return workflow.connections


@router.post("/workflows/{workflow_id}/connections", response_model=WorkflowConnectionResponse)
def create_workflow_connection(
    workflow_id: int,
    connection_data: WorkflowConnectionCreate,
    current_user: User = Depends(get_current_active_user),
    db: Session = Depends(get_db)
):
    """Create a new connection in a workflow"""
    workflow = db.query(Workflow).filter(
        Workflow.id == workflow_id,
        Workflow.owner_id == current_user.id
    ).first()
    
    if not workflow:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Workflow not found"
        )
    
    connection = WorkflowConnection(
        **connection_data.dict(),
        workflow_id=workflow_id
    )
    db.add(connection)
    db.commit()
    db.refresh(connection)
    return connection


@router.put("/workflows/{workflow_id}/connections/{connection_id}", response_model=WorkflowConnectionResponse)
def update_workflow_connection(
    workflow_id: int,
    connection_id: int,
    connection_data: WorkflowConnectionUpdate,
    current_user: User = Depends(get_current_active_user),
    db: Session = Depends(get_db)
):
    """Update a workflow connection"""
    connection = db.query(WorkflowConnection).join(Workflow).filter(
        WorkflowConnection.id == connection_id,
        Workflow.id == workflow_id,
        Workflow.owner_id == current_user.id
    ).first()
    
    if not connection:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Connection not found"
        )
    
    for field, value in connection_data.dict(exclude_unset=True).items():
        setattr(connection, field, value)
    
    db.commit()
    db.refresh(connection)
    return connection


@router.delete("/workflows/{workflow_id}/connections/{connection_id}")
def delete_workflow_connection(
    workflow_id: int,
    connection_id: int,
    current_user: User = Depends(get_current_active_user),
    db: Session = Depends(get_db)
):
    """Delete a workflow connection"""
    connection = db.query(WorkflowConnection).join(Workflow).filter(
        WorkflowConnection.id == connection_id,
        Workflow.id == workflow_id,
        Workflow.owner_id == current_user.id
    ).first()
    
    if not connection:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Connection not found"
        )
    
    db.delete(connection)
    db.commit()
    return {"message": "Connection deleted successfully"} 

File: frontend/src/stores/workflows.js
import { defineStore } from 'pinia'
import { ref, computed } from 'vue'
import api from '@/services/api'

export const useWorkflowsStore = defineStore('workflows', () => {
  const workflows = ref([])
  const currentWorkflow = ref(null)
  const loading = ref(false)
  const error = ref(null)

  const activeWorkflows = computed(() => 
    workflows.value.filter(w => w.is_active)
  )

  const fetchWorkflows = async () => {
    loading.value = true
    error.value = null
    try {
      const response = await api.get('/workflows')
      workflows.value = response.data
    } catch (err) {
      error.value = err.message
      throw err
    } finally {
      loading.value = false
    }
  }

  const fetchWorkflow = async (id) => {
    loading.value = true
    error.value = null
    try {
      const response = await api.get(`/workflows/${id}`)
      currentWorkflow.value = response.data
      return response.data
    } catch (err) {
      error.value = err.message
      throw err
    } finally {
      loading.value = false
    }
  }

  const createWorkflow = async (workflowData) => {
    loading.value = true
    error.value = null
    try {
      const response = await api.post('/workflows', workflowData)
      workflows.value.push(response.data)
      return response.data
    } catch (err) {
      error.value = err.message
      throw err
    } finally {
      loading.value = false
    }
  }

  const updateWorkflow = async (id, workflowData) => {
    loading.value = true
    error.value = null
    try {
      const response = await api.put(`/workflows/${id}`, workflowData)
      const index = workflows.value.findIndex(w => w.id === id)
      if (index !== -1) {
        workflows.value[index] = response.data
      }
      if (currentWorkflow.value?.id === id) {
        currentWorkflow.value = response.data
      }
      return response.data
    } catch (err) {
      error.value = err.message
      throw err
    } finally {
      loading.value = false
    }
  }

  const updateWorkflowBulk = async (id, workflowData) => {
    loading.value = true
    error.value = null
    try {
      const response = await api.put(`/workflows/${id}/bulk`, workflowData)
      const index = workflows.value.findIndex(w => w.id === id)
      if (index !== -1) {
        workflows.value[index] = response.data
      }
      if (currentWorkflow.value?.id === id) {
        currentWorkflow.value = response.data
      }
      return response.data
    } catch (err) {
      error.value = err.message
      throw err
    } finally {
      loading.value = false
    }
  }

  const deleteWorkflow = async (id) => {
    loading.value = true
    error.value = null
    try {
      await api.delete(`/workflows/${id}`)
      workflows.value = workflows.value.filter(w => w.id !== id)
      if (currentWorkflow.value?.id === id) {
        currentWorkflow.value = null
      }
    } catch (err) {
      error.value = err.message
      throw err
    } finally {
      loading.value = false
    }
  }

  const executeWorkflow = async (id, executionData = {}) => {
    loading.value = true
    error.value = null
    try {
      const response = await api.post(`/workflows/${id}/execute`, executionData)
      return response.data
    } catch (err) {
      error.value = err.message
      throw err
    } finally {
      loading.value = false
    }
  }

  const testWorkflow = async (id, executionData = {}) => {
    loading.value = true
    error.value = null
    try {
      const response = await api.post(`/workflows/${id}/test`, executionData)
      return response.data
    } catch (err) {
      error.value = err.message
      throw err
    } finally {
      loading.value = false
    }
  }

  const createNode = async (workflowId, nodeData) => {
    try {
      const response = await api.post(`/workflows/${workflowId}/nodes`, nodeData)
      if (currentWorkflow.value?.id === workflowId) {
        currentWorkflow.value.nodes.push(response.data)
      }
      return response.data
    } catch (err) {
      error.value = err.message
      throw err
    }
  }

  const updateNode = async (workflowId, nodeId, nodeData) => {
    try {
      const response = await api.put(`/workflows/${workflowId}/nodes/${nodeId}`, nodeData)
      if (currentWorkflow.value?.id === workflowId) {
        const index = currentWorkflow.value.nodes.findIndex(n => n.id === nodeId)
        if (index !== -1) {
          currentWorkflow.value.nodes[index] = response.data
        }
      }
      return response.data
    } catch (err) {
      error.value = err.message
      throw err
    }
  }

  const deleteNode = async (workflowId, nodeId) => {
    try {
      await api.delete(`/workflows/${workflowId}/nodes/${nodeId}`)
      if (currentWorkflow.value?.id === workflowId) {
        currentWorkflow.value.nodes = currentWorkflow.value.nodes.filter(n => n.id !== nodeId)
      }
    } catch (err) {
      error.value = err.message
      throw err
    }
  }

  const createConnection = async (workflowId, connectionData) => {
    try {
      const response = await api.post(`/workflows/${workflowId}/connections`, connectionData)
      if (currentWorkflow.value?.id === workflowId) {
        currentWorkflow.value.connections.push(response.data)
      }
      return response.data
    } catch (err) {
      error.value = err.message
      throw err
    }
  }

  const updateConnection = async (workflowId, connectionId, connectionData) => {
    try {
      const response = await api.put(`/workflows/${workflowId}/connections/${connectionId}`, connectionData)
      if (currentWorkflow.value?.id === workflowId) {
        const index = currentWorkflow.value.connections.findIndex(c => c.id === connectionId)
        if (index !== -1) {
          currentWorkflow.value.connections[index] = response.data
        }
      }
      return response.data
    } catch (err) {
      error.value = err.message
      throw err
    }
  }

  const deleteConnection = async (workflowId, connectionId) => {
    try {
      await api.delete(`/workflows/${workflowId}/connections/${connectionId}`)
      if (currentWorkflow.value?.id === workflowId) {
        currentWorkflow.value.connections = currentWorkflow.value.connections.filter(c => c.id !== connectionId)
      }
    } catch (err) {
      error.value = err.message
      throw err
    }
  }

  const clearError = () => {
    error.value = null
  }

  return {
    workflows,
    currentWorkflow,
    loading,
    error,
    activeWorkflows,
    fetchWorkflows,
    fetchWorkflow,
    createWorkflow,
    updateWorkflow,
    updateWorkflowBulk,
    deleteWorkflow,
    executeWorkflow,
    testWorkflow,
    createNode,
    updateNode,
    deleteNode,
    createConnection,
    updateConnection,
    deleteConnection,
    clearError
  }
}) 